### PR TITLE
Feature/elasticsearch multiple data node types

### LINKED ISSUE #7815: [incubator/elasticsearch] Allow heterogeneus topology for data nodes
**FEATURE REQUEST**
At the moment, we can only specify one type of data node. This is limiting in the sense that we cannot build cluster with different topologies, as we would be with hot/warm architecture (https://www.elastic.co/blog/hot-warm-architecture-in-elasticsearch-5-x).

The idea of this feature would be enabling the user to define different types of data nodes.

**Version of Helm and Kubernetes**:
Any

**Which chart**:
incubator/elasticsearch

### PR DESCRIPTION
**Pull request by rccrdpccl**
**What this PR does / why we need it**:
This PR aims at enabling the user to specify different data node types. This allows different cluster topologies that now are not possible, thus enables the possibility to implement [hot/warm architecture ](https://www.elastic.co/blog/hot-warm-architecture-in-elasticsearch-5-x)

**Which issue this PR fixes**: fixes #7815 

**Special notes for your reviewer**:
We are already running in production a very similar version of this chart, and we are starting to migrate to this version.

### CODE REVIEW

**Comment by rendhalver** [comment]
@rccrdpccl Please sign-off your commits and fix the version, It looks like your fork needs updating. There have been few PR's merged and the version is way out of date.

I also have concerns about the new pods you have added. They are not optional so it anyone upgrades there current deployment those new pods will be added and this may not be desirable.
Please make your new pods optional so we don't clobber peoples clusters.

**Comment by rccrdpccl** [comment]
@rendhalver  thanks for reviewing this quickly. The new pods will be added only if the list `nodeTypes` is not empty: `{{- range $data := .Values.dataTypes -}}`. If I should make it more explicit please let me know.

**Comment by rccrdpccl** [comment]
I've improved the readme, if it's still not clear enough please let me know!

File: incubator/elasticsearch/README.md [hunk 1]
```diff
@@ -197,3 +198,87 @@ The `tpl` function allows us to pass string values from `values.yaml` through th
...
+a specific node type (marked in nodeSelector), which will be tainted in order to avoid
+other pods being allocated in such nodes.
+
+When the `data` attribute is not declared, we will create only the data nodes declared
+in the `dataTypes` list. This is the recommended way to deploy this chart from scratch.
+
+However, if we already have deployed a version with `data` declared, we could simplify
+the migration by first adding `dataTypes`, and later removing the `data` attribute in
+another deploy.
+>>>>>>> improve README
```
Comment (rendhalver) [inline incubator/elasticsearch/README.md]: I am guessing this line might need removing.

Comment (rccrdpccl) [reply incubator/elasticsearch/README.md]: Oops...

**Comment by rccrdpccl** [comment]
@rendhalver @icereval @simonswine could you please review this PR so that we could merge it earlier? It's a bit of a pain maintaining it up-to date, as every change in data-statefulset.yaml and configmap.yaml needs to be replicated in its data-types counterpart. Thanks!

**Comment by rendhalver** [comment]
@icereval Any objections to getting this merged?

**Comment by rccrdpccl** [comment]
@simonswine @icereval I would not want to pester you, but could you please give some feedback on this PR, even if you'd like to decline it? It's quite expensive to keep up to date, and it has been hanging around here for a while now.

File: incubator/elasticsearch/README.md [hunk 2]
```diff
@@ -65,7 +65,7 @@ The following table lists the configurable parameters of the elasticsearch chart
 | `appVersion`                         | Application Version (Elasticsearch)                                 | `6.4.2`                                             |
 | `image.repository`                   | Container image name                                                | `docker.elastic.co/elasticsearch/elasticsearch-oss` |
 | `image.tag`                          | Container image tag                                                 | `6.4.2`                                             |
-| `image.pullPolicy`                   | Container pull policy                                               | `IfNotPresent`                                      |
+| `image.pullPolicy`                   | Container pull policy                                               | `Always`                                            |
```
Comment (cpanato) [inline incubator/elasticsearch/README.md]: since is using the release tag you can set the IfNotPresent.

**Comment by desaintmartin** [comment]
Something is wrong in this branch, could you clean it up?

**Comment by rendhalver** [comment]
You can back out the changes to incubator/elasticsearch
Especially this file:  stable/elasticsearch/elasticsearch-1.9.0.tgz

**Comment by rccrdpccl** [comment]
Oops, I apologise for the little mess, everything should be fixed now.

File: stable/elasticsearch/README.md [hunk 3]
```diff
@@ -64,7 +64,7 @@ The following table lists the configurable parameters of the elasticsearch chart
 | `appVersion`                         | Application Version (Elasticsearch)                                 | `6.4.2`                                             |
 | `image.repository`                   | Container image name                                                | `docker.elastic.co/elasticsearch/elasticsearch-oss` |
 | `image.tag`                          | Container image tag                                                 | `6.4.2`                                             |
-| `image.pullPolicy`                   | Container pull policy                                               | `IfNotPresent`                                      |
+| `image.pullPolicy`                   | Container pull policy                                               | `Always`                                            |
```
Comment (rendhalver) [inline stable/elasticsearch/README.md]: Why the change to 'Always'?
This should only be used when we are following a latest tag which we are not.

**Review by rendhalver — CHANGES_REQUESTED** [review]

File: stable/elasticsearch/README.md (hunk 3 above)
Comment (rendhalver) [reply stable/elasticsearch/README.md]: Actually this should reflect the default in values.

Comment (rendhalver) [reply stable/elasticsearch/README.md]: The default is `IfNotPresent` so please revert this change.

**Comment by rendhalver** [comment]
@icereval what are your thoughts on this one?

File: stable/elasticsearch/templates/data-types-configmaps.yaml [hunk 4]
```diff
@@ -0,0 +1,170 @@
+{{- $elasticsearchFullName :=  include "elasticsearch.fullname" . -}}
```
Comment (rendhalver) [inline stable/elasticsearch/templates/data-types-configmaps.yaml]: Why are you doing this?
Why don't we use the actual values instead of adding an extra layer of complexity?
It's just another thing people need to lookup to work out what you are doing.

**Review by rendhalver — CHANGES_REQUESTED** [review]

File: stable/elasticsearch/templates/data-types-configmaps.yaml (hunk 4 above)
Comment (rccrdpccl) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: If you mean "this" by assigning the value to a variable, it is because the template is referring to values and I am changing scope with the range function.

I guess another solution would be calling directly the template passing the root object, but I would have to assign the root object first, something like:
```
{{- $root := . }}
...
 labels:
    app: {{ include "elasticsearch.fullname" $root }}
```

Would it be more acceptable?

Comment (rendhalver) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: From what I can see you are adding quantifiers onto the values in your range function.
I don't see the point of defining a value that is based on another predefined variable which exists.
Why don't we just use the pre-existing variable.

Comment (rccrdpccl) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: I am afraid I am not sure I follow you: could you please give me a specific example so that I can have a better understanding of the problem you are pointing out?
I am not sure if it relates to the example I've shown below, however I could also modify all files so that they refer to root object (saved from outer scope) instead of saving multiple variables. Is this going in the direction you'd like to see?

Comment (rendhalver) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: You are defining new variables for this template based on existing variables.
You can just use the original variables. You are not saving much by redefining them.

Comment (rccrdpccl) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: Right, it makes sense. It was mainly due to my newbieness in golang templates, I did not know we could reference root by `$`.

Comment (rccrdpccl) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: Right, it makes sense. It was mainly due to my newbieness in golang templates, I did not know we could reference root by `$`.

Comment (rccrdpccl) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: Right, it makes sense. It was mainly due to my newbieness in golang templates, I did not know we could reference root by `$`.

Comment (rccrdpccl) [reply stable/elasticsearch/templates/data-types-configmaps.yaml]: Right, it makes sense. It was mainly due to my newbieness in golang templates, I did not know we could reference root by `$`.

**Comment by rccrdpccl** [comment]
@rendhalver @icereval @simonswine @cpanato @desaintmartin any other aspect that you think could be improved in this PR?

File: stable/elasticsearch/README.md [hunk 5]
```diff
@@ -106,6 +106,10 @@ The following table lists the configurable parameters of the elasticsearch chart
 | `master.antiAffinity`                | Master anti-affinity policy                                         | `soft`                                              |
 | `master.nodeAffinity`                | Master node affinity policy                                         | `{}`                                                |
 | `data.exposeHttp`                    | Expose http port 9200 on data Pods for monitoring, etc              | `false`                                             |
+| `data.enabled`                       | Enable data deployment. See `dataTypes` attribute.                  | `true`                                              |
+| `data.replicas`                      | Data node replicas (statefulset)                                    | `2`                                                 |
```
Comment (desaintmartin) [inline stable/elasticsearch/README.md]: dupe?

File: stable/elasticsearch/README.md [hunk 6]
```diff
@@ -121,6 +125,7 @@ The following table lists the configurable parameters of the elasticsearch chart
 | `data.terminationGracePeriodSeconds` | Data termination grace period (seconds)                             | `3600`                                              |
 | `data.antiAffinity`                  | Data anti-affinity policy                                           | `soft`                                              |
 | `data.nodeAffinity`                  | Data node affinity policy                                           | `{}`                                                |
+| `dataTypes`                          | Array with different data node types (see `data.` properties)       | `[]`                                                |
```
Comment (desaintmartin) [inline stable/elasticsearch/README.md]: Not clear to me, maybe state: `In case you want different data node types, optional array defining them. See documentation below.`
This PR introduce complex things and can be difficult to understand for not experienced people, we should make sure it stay simple.

**Comment by desaintmartin** [comment]
As a general note, how could we make sure the `templates/data-*` files do not diverge from the original ones?

**Comment by rccrdpccl** [comment]
> As a general note, how could we make sure the `templates/data-*` files do not diverge from the original ones?

Indeed this has been a problem for myself as well while keeping the PR up to date. I can't see an easy way to do this, unless we break BC (using for example only an list of nodes instead than specifying master, coordinating, data)

File: stable/elasticsearch/README.md (hunk 6 above)
Comment (rccrdpccl) [reply stable/elasticsearch/README.md]: I've added a description more in line with the others. Please let me know if it's still not clear

**Comment by kfox1111** [comment]
Think a default of 1 hour will be enough for the data migration? Should it be set to something huge like 10 years so that it never is crossed by default by mistake?

**Comment by desaintmartin** [comment]
@kfox1111 Well, if I'm not wrong it was already one hour before this PR. Feel free to reopen https://github.com/helm/charts/issues/8618 or a new issue to discuss it more specifically.

**Comment by rendhalver** [comment]
Can you rebase [again] to pull in the latest changes please.

**Comment by rendhalver** [comment]
Could you specify the default pods in your new data structure and use that to deploy the node rather than having two sets of resources we select between?
Does that make sense?

**Comment by rccrdpccl** [comment]
> Could you specify the default pods in your new data structure and use that to deploy the node rather than having two sets of resources we select between?
> Does that make sense?

Yes, that would make sense but however wouldn't it introduce non-BC changes? If we go down that route should we define _all_ nodes in the array, and maybe call it `nodeTypes`?

**Comment by rccrdpccl** [comment]
I am willing to discuss a solution and integrate it in the chart (re-do from scratch if needed), if we can do it swiftly - otherwise it would be time consuming to keep integrating small PRs changes within this PR.

I also would not mind if we decide to decline this PR and not integrate the feature.

What do you suggest @rendhalver @desaintmartin ?

**Comment by kfox1111** [comment]
#10319 filed
